### PR TITLE
Add register_local_source and use_generic_names funtionality to DistributedGradientTape for TF.

### PR DESCRIPTION
**Pull request by romerojosh**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description
This PR adds a couple of additional functionalities to `DistributedGradientTape` to enable better gradient handling for model parallel cases.

1. Added new method `register_local_source` to `DistributedGradientTape`: This enables users to mark a source/variable as worker local so Horovod will skip performing any global averaging on gradients associated with that variable during a call to `gradients` and will instead return the unmodified local gradient. For example:
```
    tape = hvd.DistributedGradientTape(tape)

    # Register worker local variables (i.e. local source)
    for var in model.trainable_variables:
      if <var is worker local>:
        tape.register_local_source(var)

    # Compute gradients. Any gradient associated with a var passed to register_local_source will not be modified by Horovod.
    gradients = tape.gradient(loss, model.trainable_variables)

```

2. Added a new option `use_generic_names` to the `gradients` method: In model parallel scenarios, it is often the case that common gradients on workers that need to be allreduced might be generated on different workers using different logical paths/graphs. TF naming scheme depends on the code paths that create the variables and this can lead to naming mismatches and deadlocks. This issue was resolved for Horovod operations on training variables in #2328 but not the allreduce on gradients. The solution for training variables was to use custom names and also strip the outer name scope applied by TF using the introduced `ignore_name_scope` option. The `use_generic_names` option applies the same fix to gradients by supplying custom generic names (e.g. `grad_0`, `grad_1`,...) and the `ignore_name_scope` option to the underlying allreduce calls on the gradients.

### CODE REVIEW

**Comment by MrAta** [comment]
@romerojosh we're using pretty much the same patch for our model parallel training, though we call it `PartialDistributedGradientTape` since it's a different use case than the default data parallel use case. 
What do you think about an `PartialDistributedGradientTape(tape, [local_layers])` API which is a wrapper around the `DistributedGradientTape` with local layers? I'ld be happy to raise a PR after this one is merged.

Also, regarding registering local variables, not sure if there will be use cases that need variable level granularity as opposed to layer level granularity, and hence the `[local_layers]` in  `PartialDistributedGradientTape(tape, [local_layers])`. Any thoughts on that?

**Comment by romerojosh** [comment]
Hey @MrAta, thanks for the comment!

Adding an optional list argument to `DistributedGradientTape` like `DistributedGradientTape(tape, local_vars=[...])`  was an alternative approach instead of the `register_local_variables` I considered, but thought it would be more flexible to allow users to to add variables to this list after creating the tape (perhaps in situations where the variables aren't known at tape creation time). 

I think adding a `PartialDistributedGradientTape(tape, [local_layers])` wrapper in a follow up PR would be great! In terms of using layers vs variables, at least for this PR, I wanted to use the same level of granularity that Horovod uses internally to schedule the communication operations (so, variables). For the higher level wrapper you mention, I think layers could make more sense from a user convenience standpoint.

**Comment by MrAta** [comment]
Thanks for the insight, @romerojosh. Great, I'll raise a PR for `PartialDistributedGradientTape` once this PR is merged then.

File: horovod/tensorflow/__init__.py [hunk 1]
```diff
@@ -520,7 +535,9 @@ def allreduce_grads(grads, vars=None):
                                                                op=op,
                                                                prescale_factor=prescale_factor,
                                                                postscale_factor=postscale_factor,
-                                                               process_set=process_set)
+                                                               process_set=process_set,
+                                                               name=f"grad_{i}" if use_generic_names else None,
+                                                               ignore_name_scope=True if use_generic_names else False)
```
Comment (MrAta) [inline horovod/tensorflow/__init__.py:542]: is there any reason we don't do `ignore_name_scope=use_generic_names` given that both are booleans?

Comment (romerojosh) [reply horovod/tensorflow/__init__.py]: No particular reason. Your suggestion is cleaner so I made that change. Thanks!

**Review by maxhgerlach — APPROVED** [review]
Changes look good to me and the new functionality seems very useful!

The one suggestion that came to mind would be to add a small unit test for `register_local_source()`. However, as it stands we have quite little test coverage for the various existing options of `DistributedGradientTape` and `DistributedOptimizer`, so that feels optional to me. 🙂
